# HW01
Welcome to the `assignment 1` of Introduction to Data Analysis.

In this assignment, you will practice basics of python, numpy and pandas.

Please follow the instructions below:

1. Write down your code between <br>
&nbsp;&nbsp;&nbsp;&nbsp; **\# BEGIN_YOUR_CODE**<br>
&nbsp;&nbsp;&nbsp;&nbsp; and<br>
&nbsp;&nbsp;&nbsp;&nbsp; **\# END_YOUR_CODE**.

2. Do not use **external libraries**. (i.e., Do not use any `import` in your code) <br>
   Your code will fail to execute and get **0 score** if you use them.

3. Rename this file to **[student_id].ipynb** (e.g. 20230000.ipynb) and submit it to PLMS. <br>
   There is **30% penalty** if you do not follow the submission format.

4. Submission late is not accepted.
   You will get **No score** for late submission.

In [103]:
import numpy as np
import pandas as pd

## Problem 1. Factorial [2 points]
Given `n`, implement function `factorial(n)` that calculates `n!`. <br>
Assume `n` is zero or a positive integer. Just so you know, `0!` is defined as `1`.<br>


In [104]:
def factorial(n):
    # BEGIN_YOUR_CODE
    if n in (0, 1):
      return 1
    return n * factorial(n-1)
    # END_YOUR_CODE

In [105]:
# example
for i in range(1, 10):
    print(factorial(i))
    
# expected:
# 1
# 2
# 6
# 24
# 120
# 720
# 5040
# 40320
# 362880

1
2
6
24
120
720
5040
40320
362880


### Explaination

If `n` is 0 or 1, `facorial(n)` is 0. \\
If `n` is bigger than 1, `facorial(n)` returns `n` * `factorial(n-1)`.

For example, the process of calculating `factorial(5)` is:

`factorial(5)` 

= 5 $\times$ `factorial(4)` 

= 5 $\times$ 4 $\times$ `factorial(3)` 

= 5 $\times$ 4 $\times$ 3 $\times$ `factorial(2)` 

= 5 $\times$ 4 $\times$ 3 $\times$ 2 $\times$ `factorial(1)` 

= 5 $\times$ 4 $\times$ 3 $\times$ 2 $\times$ 1 = 120


## Problem 2. Frequent Word Count [2 points]
Implement a function `freq_word_count(filename)` that takes a filename as input. <br>
This function should open the specified file and determine which words appear most frequently. It should return a `tuple` where the first item is the most frequent word and the second is the count of that word.<br>
If there are multiple words with the same maximum frequency, you can return any one of them.


In [106]:
def freq_word_count(filename):
    '''
    filename: .txt file
    '''
    # BEGIN_YOUR_CODE
    f = open(filename)
    lines = f.readlines()
    word_list = dict()
    for line in lines:
        words = line.split()
        for word in words:
            word.strip()
            if word in word_list:
                word_list[word] = word_list[word] + 1
            else:
                word_list[word] = 1
    max_word = 0
    count = 0
    for word in word_list:
        if(word_list[word] > count):
            max_word = word
            count = word_list[word]
    return (max_word, count)
    # END_YOUR_CODE

In [120]:
# example
# --- test.txt
# a b c d
# a
# a
# cdcd
# --- EoF
filename = 'test.txt'
f = open(filename, 'w')
f.write("a b c d\na\na\ncdcd")
f.close()
print(freq_word_count(filename))

# expected: 
# ('a', 3)

('a', 3)


### Explaination

`f` reads the file named as `filename`, then each lines are stored in `lines` in the form of tuple. `word_list` is a dictionary with key and value, and its key is word, and value is the number of each word. This function counts all words from `lines`, records in the `word_list`, then it choosees the most frequently occured word in `word_list`.

This function words as followes:

let file is "a b c d `\n` a `\n` a `\n` cdcd" where `\n` is newline. 

Then lines is `["a b c d \n", "a \n", "a \n", "cdcd"]`. 

word is `a`, `b`, `c`, `d`, `a`, `a`, `cdcd`. 


So, the most frequent word is `a`, 3 times.

## Problem 3. Median Score [3 points]
Given `students`, implement function `median_score(students)` that returns a dictionary with each student_id as a key and the median score as the value.<br> If the number of scores for a student is odd, the median is the middle score. If the number of scores is even, the median is the average of the two middle scores.

In [108]:
class Student:
    def __init__(self, args):
        name, student_id, grade =  args
        self.name = name
        self.student_id = student_id
        self.grade = grade

def median_score(students):
    '''
    Args:
    students (dict): list of Student instances

    Returns:
    result (dict): A dictionary where each key is a student_id and the value is the median score.
    '''
    # BEGIN_YOUR_CODE
    students = sorted(students, key = lambda Student: Student.grade)
    median = 0
    median_name = ""
    length_of_list = len(students)
    if(length_of_list % 2 == 0):
        median_name = students[length_of_list // 2 - 1].name
        median = (students[length_of_list // 2 - 1].grade + students[length_of_list // 2].grade) / 2
    else:
        median_name = students[length_of_list // 2].name
        median = students[length_of_list // 2].grade
    result = {median_name: median}
    # END_YOUR_CODE
    return result

In [109]:
# example
# When the number of students is odd.
t1 = [Student(["KIM", 20230001, 10]), Student(["HUH", 20230002, 30]), Student(["LEE", 20230003, 20])]
print(median_score(t1))
# expected: {'LEE': 20}

#When the number of students is even.
t2 = [Student(["KIM", 20230001, 10]), Student(["HUH", 20230002, 30]), Student(["LEE", 20230003, 20]), Student(["PARK", 20230004, 15])]
print(median_score(t2))
# expected: {'PARK': 17.5}

{'LEE': 20}
{'PARK': 17.5}


## Explanation

The argument is a list of `Student` instances. To get the median value, the function does as:

1. Sort the elements of the list with key is grade (ascending order)

2. the return, dictionary, is determined as:

2-1. If the number of elements is odd, student's name which is placed in middle one and the value is his(her) grading.

2-2. If the number of elements is even, the key is the student to the left of the center two students, and the value is the arithmetic mean of the center two students.


## Problem 4. Vector Norm [2 points]
Given two numpy arrays arr1 and arr2, and an integer n, implement a function `vector_norm` that calculates the n-th vector norm of the difference between arr1 and arr2. <br>
You need to utilize the numpy library for this problem.

In [110]:
def vector_norm(arr1, arr2, n):
    # BEGIN_YOUR_CODE
    arr = arr1 - arr2
    norm = 0
    for elem in arr:
        norm += (elem ** 2)
    return norm**(1/2)
    # END_YOUR_CODE

In [111]:
# example
a1 = np.array([1, 2, 3, 4])
a2 = np.array([4, 1, 6, 8])
print("arr: ", a1 - a2)
print(vector_norm(a1, a2, 4))

# expected:
# arr:  [-3  1 -3 -4]
# 5.916079783099616

arr:  [-3  1 -3 -4]
5.916079783099616


## Explanation

The norm of N-vector $\vec v$ is defined as:
$$\|v\| = \sqrt{\sum_i^N v_i^2} = \sqrt{v_1^2 + v_2^2 + \cdots + v_n^2}$$

The difference of two vector can be calculated element-wisely.

For example, $(1, 2, 3, 4)^T - (4, 1, 6, 8)^T = (1-4, 2-1, 3-6, 4-8)^T = (-3, 1, -3, 4)^T$

Note
* The order of subtraction does not affect the result of the norm of a vector.

## Problem 5. CSV Modification [5 points]
Your goal is to modify given csv file with below constraints. <br>
The inputs are paths of the original data and modified data. <br>
You need to utilize pandas library for this problem.

### Constraints
- The requirements must be followed in the same order as given below.<br>
  (If not, you might attain different results although you followed everything correctly.)
1. The modified csv file should contain rows where "Active" cases are greater than 10,000.
2. The modified csv file should only have `Europe` region.
3. Add new columns called "Mortality_Rate" and "Recovery_Rate", which are calculated as `(Deaths / Confirmed) * 100`, `(Recovered / Confirmed) * 100`.
4. Sort the data by "Mortality Rate (%)" in descending order


In [112]:
import pandas as pd


def covid(original_file, modified_file):
    df = pd.read_csv(original_file)
    # BEGIN_YOUR_CODE
    df = df[df['Active'] > 10000]
    df = df[df["WHO Region"] == "Europe"]
    df["Mortality_Rate"] = (df["Deaths"] / df["Confirmed"]) * 100
    df["Recovery_Rate"] = (df["Recovered"] / df["Confirmed"]) * 100
    df = df.sort_values("Mortality_Rate", ascending = False)
    # END_YOUR_CODE
    df.to_csv(modified_file, index=False)
    return df

In [113]:
# example
# The csv file is placed in root workspace folder   
covid("COVID.csv", "COVID_revised.csv")

Country/Region  Confirmed  Deaths  Recovered  Active  New cases  \
177  United Kingdom     301708   45844       1437  254427        688   
16          Belgium      66428    9822      17452   39154        402   
85            Italy     246286   35112     198593   12581        168   
61           France     220352   30212      81212  108928       2551   
120     Netherlands      53413    6160        189   47064        419   
157           Spain     272421   28432     150376   93613          0   
161          Sweden      79395    5700          0   73695        398   
137         Romania      45902    2206      25794   17902       1104   
93       Kyrgyzstan      33296    1301      21205   10790        483   
135        Portugal      50299    1719      35375   13205        135   
172          Turkey     227019    5630     210469   10920        919   
175         Ukraine      67096    1636      37202   28258        835   
147          Serbia      24141     543          0   23598        411   
7           Armenia      37390     711      26665   10014         73   
138          Russia     816680   13334     602249  201097       5607   
84           Israel      63985     474      27133   36378       2029   
89       Kazakhstan      84648     585      54404   29659       1526   

     New deaths  New recovered  Deaths / 100 Recovered  Confirmed last week  \
177           7              3                 3190.26               296944   
16            1             14                   56.28                64094   
85            5            147                   17.68               244624   
61           17            267                   37.20               214023   
120           1              0                 3259.26                52132   
157           0              0                   18.91               264836   
161           3              0                     inf                78048   
137          19            151                    8.55                38139   
93           24            817                    6.14                27143   
135           2            158                    4.86                48771   
172          17            982                    2.67               220572   
175          11            317                    4.40                60767   
147           9              0                     inf                21253   
7             6            187                    2.67                34981   
138          85           3077                    2.21               776212   
84            4            108                    1.75                52003   
89            0           1833                    1.08                73468   

     1 week change  1 week % increase WHO Region  Mortality_Rate  \
177           4764               1.60     Europe       15.194824   
16            2334               3.64     Europe       14.785934   
85            1662               0.68     Europe       14.256596   
61            6329               2.96     Europe       13.710790   
120           1281               2.46     Europe       11.532773   
157           7585               2.86     Europe       10.436787   
161           1347               1.73     Europe        7.179293   
137           7763              20.35     Europe        4.805891   
93            6153              22.67     Europe        3.907376   
135           1528               3.13     Europe        3.417563   
172           6447               2.92     Europe        2.479969   
175           6329              10.42     Europe        2.438297   
147           2888              13.59     Europe        2.249285   
7             2409               6.89     Europe        1.901578   
138          40468               5.21     Europe        1.632708   
84           11982              23.04     Europe        0.740799   
89           11180              15.22     Europe        0.691097   

     Recovery_Rate  
177       0.476288  
16       26.272054

## Explanation
Each lines does as:
1. Select the columns which `Active` is bigger than 10000
2. Select the columns which `WHO Region` is Europe
3. Add Mortality_Rate
4. Add Recovery_Rate
5. Sort in descending order based on Mortality_rate.

## Problem 6. Employee and Department [6 points]
For this problem, three csv files, `departments.csv`, `employees.csv` and `employees2.csv`, are given. <br>
There are 2 sub problems. <br>
You need to utilize pandas library for this problem.

### 6.a Employee Table [3 points]
Make employee table that has `name`, `salary` and `department_name` as columns. <br>
Note that each department has its own `department_id` and `department_name`.

In [114]:
def emp_table(dep, emp1, emp2):
    # BEGIN_YOUR_CODE
    ddf = pd.read_csv(dep)
    edf1 = pd.read_csv(emp1)
    edf2 = pd.read_csv(emp2)
    df = pd.concat([edf1, edf2])
    df = pd.merge(df, ddf, on = 'department_id', how = 'inner')
    df = df[['name', 'salary', 'department_name']]
    # END_YOUR_CODE
    return df

In [115]:
# example
emp_table("departments.csv", "employees.csv", "employees2.csv")

name  salary department_name
0    John    5000           sales
1   Allen    6000      accounting
2  Martin    3500        research
3    Mary    5500           sales
4   Smith    4500        research
5   Scott    7100      accounting
6     Tom    4500           sales
7   Admas    5000        research
8  Hunter    3000        research
9    King    5100      accounting

### 6.b Highest Average Salary [3 points]
Find the department that has the highest average salary.<br>
The output should be a dictionary with the `department_name` as the key and its highest average salary as the value. <br>
You can use the `emp_table` provided in 6.a.

In [116]:
def highest_avg_salary(dep, emp1, emp2):
    # BEGIN_YOUR_CODE
    avg_salary = emp_table(dep, emp1, emp2).groupby('department_name')['salary'].mean() 
    highest_avg_dept = avg_salary.idxmax()
    highest_avg_salary = avg_salary.max()
    result = dict()
    result[highest_avg_dept] = highest_avg_salary.item()
    return result
    # END_YOUR_CODE

In [117]:
# example
highest_avg_salary("departments.csv", "employees.csv", "employees2.csv")

# expected:
# {'accounting': 6066.666666666667}

{'accounting': 6066.666666666667}

## Explanation
### 6.a
`emp_table()` function gets three arguments: the name of departments, employees, employees2 files.

First, the function reads three csv files as pandas dataframe.

Second, it connects two employee dataframe using `concat()`.

Third, it does innerjoin from employee to departments using `merge()`

Fourth, it selects columns `name`, `salary`, and `department_name`

### 6.b
`highest_avg_salary()` function gets three arguments: the name of departments, employees, employees2 files.

First, the function gets the average salary of each department using 6.a `emp_table()` and `groupby()`. The average salary dataframe is like as:

department_name<br>
accounting    6066.666667<br>
research      4000.000000<br>
sales         5000.000000<br>
Name: salary, dtype: float64

Second, it select the name and salary of department which has the most average salary using `idxmax()` and `max()`.

Third, it returns the dictionary which it's key is department name and value is salary.

Because the data type of average calculated with `groupby()` is `np.float32`, we have to use `.item()` to convert from `np.float32` to float.